README.md
# COLDSALT

Coldsalt is a simple Python program to speed up ingesting API data for penetration testing.

## Getting Started

Written and tested with Python 2.7.10. Not tested under Python 3.

### Prerequisites

Coldsalt uses the Requests library.   

To see if you have Requests installed, use the following command (assuming you use pip):  
```shell
pip show requests
```

To install Requests with pip, use the following command:  
```shell
pip install requests
```

### Optional functionality

Coldsalt has hooks to use BurpBuddy to automatically populate Intruder and Repeater.

You can get BurpBuddy here: https://github.com/tomsteele/burpbuddy/releases  

Coldsalt was developed and tested with BurpBuddy v3.0.0-BETA.

## Running Coldsalt

Usage example:
```
python coldsalt.py --mode swagger -i swagger_example.json --checkonly
python coldsalt.py --mode postman -i postman_example.json -e environment.json --output awesome.csv

-h, --help            show this help message and exit
-a, --allmethods      Make API calls for all HTTP methods including state
                      change related methods.
--burpbuddy BURPBUDDY
                      IP:port of Burp Budy instance. Defaults to
                      localhost:8081.
--checkonly           enable debug mode. No requests will be sent
-c CONTENTTYPE, --contenttype CONTENTTYPE
                      Force or override Content-Type on requests
-e ENVIRONMENT, --environment ENVIRONMENT
                      The environment.json file to parse.
--headers HEADERS     The headers.json file to parse.
-i INPUT, --input INPUT
                      The JSON collection file, Postman or Swagger, to
                      parse.
--mode {curl,postman,swagger}
                      Which mode, or collection type, to use. Choices are
                      'curl', 'postman', or 'swagger'
--output OUTPUT       write output/audit log to a csv
--proxy PROXY         IP:Port of proxy to use. Defaults to localhost:8080.
-v, --verbose         Enable verbose output.
-x, --xperimental     enable Xperimental Burp mode

```

By default, Coldcalt will only send GET, HEAD, and OPTION requests. You can override this with the -a and --allmethods flags.

If you're unsure about the quality of the documentation provided to you by the project team, start with the --checkonly flag. This will run the "parser" "engine" and look for missing parameters.  

If you do not have an environment file, COLDSALT will create a file stub for you. You can copypasta the output into your file editor of choice. Be sure to clean up errant line breaks and update the values.

Even if you have an environment file, if there are placeholders/variables defined in your collection file that are not defined in your environment file, COLDSALT will still output templated key value pairs for the missing parameters.

Once all the things check out, make sure you have Burp running on 8080 and run the script again.

### Supported Formats

 - Postman v2
 - Swagger
 - curl

If you are using a Postman file, make sure it is Postman v2.

To use curl requests, make sure they are in a text file, one request per line with the host at the end of the command. Generally, you can get a file ready with a combo of grep and awk.

### Setting Headers

Suppose you have a Swagger or Postman file that doesn't have any placeholders for required headers, what then?

Coldsalt gives you two options.  
``` -c CONTENTTYPE, --contenttype CONTENTTYPE ``` can be used to set an arbitrary Content-Type header.  
``` --headers HEADERS ``` lets you specify a json file that can contain any number of headers in the following format:
```
{"name": "generic header file",
"values": [{"key": "foo", "value": "bar"},
	{"key": "X-Test", "value": "true"}
	]}
```

Headers specified with ```--headers``` will override any duplicated headers from within the Swagger or Postman files.  
Content-Type set by ```-c``` will override all other Content-Type values.  

## Contributing

Please see [Contributing](/CONTRIBUTING.md).

## License

This project is licensed under the Apache License v2. Please see [License](/LICENSE) for more details.


CHANGELOG.md
# Changelog

## v0.7.1  

### New Features  
- Added some robustness to the way variables are handled with Curl requests with JSON data. It's still not perfect, but it's better.    


## v0.7.0  

This version is a Release Candidate for version 1.0. While more testing is needed, this version should support all methods and body contents for Curl, Postman, and Swagger.  

### New Features  
- Enhanced support for Curl with data elements. `-d` and `--data` should both parse correctly now. JSON and form encoded should parse as well. Placeholders can be injected with variables as well.  
- Minor text changes on ERROR messages.   
- Cleaned up a stray print statement.  


## v0.6.0  

### New Features  
- Added initial support for request bodies for Curl. Initial code supports JSON specified with -d.  

### Known Issues  
- Other forms of data, such as --data or form encoded, won't parse right now. More to come. TBD.
- Curl parser might not inject variables into request bodies yet. I should probably test if it does. It might. But it might not. No promises.   


## v0.5.1

### Fixes  
- Added some error handling for parsing Swagger files without Summary fields.


## v0.5.0

- Added a new flag ```--headers``` to allow for arbitrary headers added to all requests.


## v0.4.1

### Fixes
- Fixed Swagger file parsing when body parameters are inline objects instead of references.


## v0.4

## New Features
- Added a new flag ```-c``` and ```--contenttype``` to force or override the content-type header sent on requests.

### Fixes
- When parsing body data for Swagger, the resulting string wasn't proper JSON. This has been fixed.  


## v0.3

### New Features  
- POST, PUT, and PATCH bodies should be populated from Swagger collections. A placeholder/variable is injected into the body for each parameter. The placeholder gets parsed before the request is sent just like any other parameter.


## v0.2

### New Features  
- Initial support of All Methods option. See below.  
- POST, PUT, and PATCH bodies should be populated from Postman collections.  

### Fixes
- Proxy strings should be concatenated correctly now.
- Parsing Postman files should be more robust.

### Known Issues  
- POST, PUT, and PATCH bodies from Swagger and Curl are not currently populated.  


cs_parser_curl.py
####
## CS_Parser_Curl
## Simple helper type class for parsing curl input into ColdSalt internal format
## Author: jason nordenstam
####

import json

class CSParserCurl:

    def __init__(self, verbose = False):
        self.verbose = verbose


    def parseCurl(self, inputfile):
        endpoint_collection = []
        lines = []
        print "\nParsing curl file %s" % str(inputfile)
        with open(inputfile) as f:
            lines = f.readlines()

        if self.verbose:
            print "Found %s lines..." % len(lines)

        for line in lines:
            if line.find("\'") > 0:
                line = line.replace("\'", "\"")

            api_endpoint = {}
            # method
            method_index = line.find("-X")
            # if a method is not specified, treat it as a GET
            if method_index == -1:
                method = "GET"
            else:
                method = line[method_index : method_index + 16].split()[1]
            api_endpoint['method'] =  method

            # headers
            headers = {}
            headers_index = line.find("-H")
            # each header starts with -H so look through them
            for x in xrange(line.count("-H")):
                # split out the current header
                header = line.split("-H")[x+1].split("\"")[1]
                # jam it into the dictionary
                headers[header.split(":")[0]] = header.split(":")[1].strip()
            # add the dictionary to the endpoint object
            api_endpoint['headers'] = headers
            # url stuff
            if line.find("\"http") == -1:
                # maybe it's not quoted?
                url = line[line.find("http"):].strip()
            else:
                url = line[line.find("\"http"):].split("\"")[1].strip()

            api_endpoint['url'] = url
            # name - curl doesn't really have a place for names, so grab the stuff
            # after /v# and use that
            base_name = url[url.find("/v"):url.find("?")].strip()
            name = method + " " + base_name

            api_endpoint['name'] = name

            # TODO this really only supports json data with -d
            # TODO make this handle --data and non-jsaon data
            # data - could be -d or --data
            brk_str = ""
            if "--data" in line:
                 brk_str = "data"
            elif "-d" in line:
                brk_str = "d"

            # if no data, do nothing
            if brk_str != "":
                tmp = line.split("-")
                for x in tmp:
                    if x.startswith(brk_str):
                        # slice out the data, add the length of the brk_str since find will
                        # return the start index of our string
                        tmp_data = x[x.find(brk_str)+len(brk_str):].strip()
                        # need to clean up opening and closing quotes
                        # determine if JSON or form encoded
                        #print tmp_data
                        if tmp_data[1:len(tmp_data)-1].startswith("{"):
                            try:
                                jsn_data = json.loads(tmp_data[1:len(tmp_data)-1])
                            except ValueError as e:
                                print "[!] Error occured parsing json. Please verify format is correct."
                                jsn_data = ""

                            api_endpoint["body"] = jsn_data
                        else:
                            api_endpoint["body"] = tmp_data[1:len(tmp_data)-1]
            else:
                api_endpoint["body"] = ""

            endpoint_collection.append(api_endpoint)

            if(self.verbose):
                print "    " +  api_endpoint["name"]
                print "      " + api_endpoint["url"]
                print "      " + api_endpoint["method"]
                print "      " + str(api_endpoint["body"])
                print "      ----------------"

        return endpoint_collection


cs_parser_postman.py
####
## CS_Parser_Postman
## Simple helper type class for parsing Postman files into ColdSalt internal format
## Author: jason nordenstam
####

import sys
import json


class CSParserPostman:

    def __init__(self, verbose = False):
        self.verbose = verbose


    def parsePostman(self, inputfile):
        endpoint_collection = []
        print "\nParsing postman collection file %s" % str(inputfile)
        postman_data = json.loads(open(inputfile).read())
        try:
            postman_items = postman_data["item"]
        except KeyError as ke:
            print "Key Error in postman file. Is it v2?"
            sys.exit("Invalid Postman File")

        if(self.verbose):
            print postman_data["info"]["name"]

        for endpoints in postman_items:

            if "item" not in endpoints:
                api_endpoint = {}
                headers = {}
                api_endpoint['name'] = endpoints["name"]
                if "raw" in endpoints["request"]["url"]:
                    api_endpoint['url'] = endpoints["request"]["url"]["raw"]
                else:
                    api_endpoint['url'] = endpoints["request"]["url"]
                api_endpoint['method'] =  endpoints["request"]["method"]
                # get headers
                for value in endpoints["request"]["header"]:
                    headers[value["key"]] = value["value"]

                if "body" in endpoints["request"]:
                    if "raw" == endpoints["request"]["body"]["mode"]:
                        api_endpoint["body"] = endpoints["request"]["body"]["raw"]
                    else:
                        api_endpoint["body"] = ""
                else:
                    api_endpoint["body"] = ""

                api_endpoint['headers'] = headers
                # once the data for an endpoint is parsed out, add it to dictionary
                endpoint_collection.append(api_endpoint)
                if(self.verbose):
                    print "  " +  endpoints["name"]
                    print "    " +  api_endpoint["name"]
                    print "      " + str(api_endpoint["url"])
                    print "      " + api_endpoint["method"]
                    print "      " + api_endpoint["body"]
                    print "      ----------------"
            else:
                for items in endpoints["item"]:

                    api_endpoint = {}
                    headers = {}
                    api_endpoint['name'] = items["name"]
                    api_endpoint['url'] = items["request"]["url"]
                    api_endpoint['method'] =  items["request"]["method"]
                    # get headers
                    for value in items["request"]["header"]:
                        headers[value["key"]] = value["value"]

                    api_endpoint['headers'] = headers
                    # once the data for an endpoint is parsed out, add it to dictionary
                    if "body" in endpoints["request"]:
                        if "raw" == endpoints["request"]["body"]["mode"]:
                            api_endpoint["body"] = endpoints["request"]["body"]["raw"]
                        else:
                            api_endpoint["body"] = ""
                    else:
                        api_endpoint["body"] = ""

                    endpoint_collection.append(api_endpoint)
                    if(self.verbose):
                        print "  " +  endpoints["name"]
                        print "    " +  api_endpoint["name"]
                        print "      " + api_endpoint["url"]
                        print "      " + api_endpoint["method"]
                        print "      ----------------"

        return endpoint_collection


cs_parser_swagger.py
####
## CS_Parser_Swagger
## Simple helper type class for parsing Swagger files into ColdSalt internal format
## Author: jason nordenstam
####

import sys
import json

standard_methods = ["GET", "HEAD", "OPTIONS", "PUT", "POST", "PATCH", "DELETE", "OPTIONS", "TRACE"]


class CSParserSwagger:

    def __init__(self, verbose = False):
        self.verbose = verbose

    def pickEnvironment(self, endpoints):
        print "\nMultiple environment endpoints detected. Please Select one."
        for endpoint in endpoints:
            print "%s - %s: %s: %s" % (endpoint['count'], endpoint['accessibility'], endpoint['type'], endpoint['url'])
        selection = int(raw_input("Select: "))

        return selection

    def parseSwagger(self, inputfile):
        endpoint_collection = []
        print "\nParsing Swagger collection file %s" % str(inputfile)
        swagger_data = json.loads(open(inputfile).read())
        swagger_items = swagger_data["paths"]
        if(self.verbose):
            print swagger_data["info"]["title"]

        # swagger can define URLs in a few different ways
        # first let's check for the "endpoints" definitions
        if "x-api-definition" in swagger_data:
            if "endpoints" in swagger_data["x-api-definition"]:
                # even with "endpoints" there are external and internal options
                endpoints = []
                endpoint_count = 0
                for accessibility in swagger_data["x-api-definition"]["endpoints"].keys():
                    for endpoint_key in  swagger_data["x-api-definition"]["endpoints"][accessibility].keys() :
                        if swagger_data["x-api-definition"]["endpoints"][accessibility][endpoint_key] is not None:
                            point = {}
                            point['count'] = endpoint_count
                            point['accessibility'] = accessibility
                            point['type'] = endpoint_key
                            point['url'] = swagger_data["x-api-definition"]["endpoints"][accessibility][endpoint_key]
                            endpoints.append(point)
                            endpoint_count = endpoint_count + 1
                _error_count = 0
                # if there are mulitple environments, prompt the user to pick one
                _selection = self.pickEnvironment(endpoints)
                # check and see if it's in range
                while _selection > endpoint_count - 1 or _selection < 0:
                    _error_count = _error_count + 1
                    print "Invalid selection. Pick an environment between 0 and %s" % (endpoint_count-1)
                    _selection = self.pickEnvironment(endpoints)
                    if _error_count > 5:
                        print "Stop messing around."
                        sys.exit(1)

                base_url = endpoints[_selection]["url"]
        else:
            #if there aren't endpoints, try to use host and basePath
            # but also check for a scheme
            if "schemes" in swagger_data:
                base_url = swaggers_data["schemes"] + swagger_data["host"] + swagger_data["basePath"]
            else:
                base_url = "http://" + swagger_data["host"] + swagger_data["basePath"]

        if(self.verbose):
            print "base_url= " + base_url

        # swagger can define parameters in several ways so...
        # first we'll check for 'global' parameters
        global_params = {}
        try:
            if swagger_data["parameters"] is not None:
                for param in swagger_data["parameters"].keys():
                    #print param
                    _local_param = {}
                    _local_param["name"] = swagger_data["parameters"][param]["name"]
                    _local_param["in"] = swagger_data["parameters"][param]["in"]

                    global_params[swagger_data["parameters"][param]["name"]] = _local_param
        except KeyError:
            print "No global parameters found."

        for path_key in swagger_items.keys():
            for method_key in swagger_items[path_key].keys():
                if method_key.upper() in standard_methods:
                    api_endpoint = {}
                    headers = {}
                    queryString = {}

                    try:
                        api_endpoint['name'] = swagger_items[path_key][method_key]["summary"]
                    except KeyError:
                        api_endpoint['name'] = "%s %s" % (method_key, path_key)

                    api_endpoint['method'] = method_key.upper()
                    api_endpoint['body'] = ""
                    # handle parameters
                    # swagger parameters can be defined in several ways
                    # first check if they are defined under
                    if "parameters" in swagger_items[path_key][method_key].keys() :
                        for param in swagger_items[path_key][method_key]["parameters"]:
                            _local_param = param
                            #print str(_local_param)
                            if "$ref" in param:
                                # if the param references a global param, grab that param
                                # and 'overwrite' the 'loop' 'copy'
                                param_ref = param["$ref"].split("#/parameters/")[1]
                                _local_param = global_params[param_ref]

                            if(_local_param["in"] == "header"):
                                headers[str(_local_param["name"])] = str("{" + _local_param["name"] + "}")
                                global_params[str(_local_param["name"])] = str(_local_param)

                            if(_local_param["in"] == "query"):
                                queryString[_local_param["name"]] = "{" + _local_param["name"] + "}"
                                global_params[_local_param["name"]] = _local_param

                            if(_local_param["in"] == "body"):
                                # get the ref....
                                _body = {}

                                #print _local_param["in"]
                                # check for a schema references
                                if "$ref" in _local_param["schema"].keys():
                                    _raw_body = swagger_data["definitions"][_local_param["schema"]["$ref"].split("#/definitions/")[1]]["properties"]
                                    for _body_item in _raw_body.keys():
                                        _body[str(_body_item)] = "{%s}" % (str(_body_item))
                                else:
                                    # if there is no reference, parse the 'raw' body
                                    if "type" in _local_param["schema"].keys():
                                        if "object" == _local_param["schema"]["type"]:
                                            _raw_body =_local_param["schema"]["properties"]
                                            for _body_item in _raw_body.keys():
                                                _body[str(_body_item)] = "{%s}" % (str(_body_item))

                                api_endpoint["body"] = str(_body)
                    # format querysting into url
                    if(bool(queryString)):
                        formatted= "&".join("%s=%s" % (key, queryString[key]) for key in queryString.keys())
                        api_endpoint['url'] = base_url + path_key + "?" + formatted
                    else:
                        api_endpoint['url'] = base_url + path_key
                    api_endpoint['headers'] = headers
                    # add to dictionary
                    endpoint_collection.append(api_endpoint)
                    if(self.verbose):
                        print "    " +  api_endpoint["name"]
                        print "      " + api_endpoint["url"]
                        print "      " + api_endpoint["method"]
                        print "      " + str(api_endpoint["body"])
                        print "      ----------------"
        return endpoint_collection


coldsalt.py
import sys, json, argparse, requests, re, random, csv
from cs_parser_postman import CSParserPostman    # used to parse Postman
from cs_parser_swagger import CSParserSwagger    # used to parse swagger input
from cs_parser_curl import CSParserCurl    # Used to parse curl input
from requests.packages.urllib3.exceptions import InsecureRequestWarning


def parseURL(urlString):
    """Parses a request URL and replaces all placeholders/variables with the associated values from the environment file."""
    finalUrl = urlString
    patterns = {"postman": "{{[\w]*}}*", "swagger": "{[\w]*}*", "curl": "{[\w]*}*"}
    parameters = re.findall(patterns[args.mode], urlString)

    for param in parameters:
        cleanParam = param.replace("{","").replace("}","")
        if cleanParam not in env_var:
            print "[!] Found parameter not in ENVIRONMENT VARIABLES: " + str(cleanParam)
            if cleanParam not in missing_parameters:
                missing_parameters.append(cleanParam)
        finalUrl = finalUrl.replace(param, env_var.get(cleanParam, ""))
    return finalUrl

def parseBody(bodyString):
    """Parses a request body and replaces all placeholders/variables with the associated values from the environment file."""

    finalBody = bodyString

    patterns = {"postman": "{{[\w]*}}*", "swagger": "{[\w]*}*", "curl": "{[\w]*}*"}

    try:
        parameters = re.findall(patterns[args.mode], bodyString)
    except TypeError:
        parameters = re.findall(patterns[args.mode], json.dumps(bodyString))

    for param in parameters:
        #print "param: " + param
        cleanParam = param.replace("{","").replace("}","")
        #print cleanParam
        if "" != cleanParam:
            if cleanParam not in env_var:
                print "[!] Found parameter not in ENVIRONMENT VARIABLES: %s !" % cleanParam
                if cleanParam not in missing_parameters:
                    missing_parameters.append(cleanParam)
            if type(finalBody) is not dict:
                finalBody = finalBody.replace(param, env_var.get(cleanParam, ""))
            else:
                # TODO this only searches two levels deep in JSON dict
                for key, value in finalBody.items():
                    if type(value) is dict:
                        for k, v in finalBody[key].items():
                            if value == param:
                                finalBody[key] == env_var.get(cleanParam)
                    if value == param:
                        finalBody[key] == env_var.get(cleanParam)

    # make sure we end up with a json string with double quote delimiters
    if type(finalBody) is not dict:
        finalBody = finalBody.replace("\'","\"")

    return finalBody




def sendRequest(requestUrl, requestMethod, requestHeaders, requestBody):
    """Send the request as passed in."""
    r = ""
    if requestMethod == "GET":
        r = requests.get(requestUrl, headers=requestHeaders, proxies=p, verify=False)
    elif requestMethod == "HEAD":
        r = requests.head(requestUrl, headers=requestHeaders, proxies=p, verify=False)
    elif requestMethod == "OPTIONS":
        r = requests.options(requestUrl, headers=requestHeaders, proxies=p, verify=False)
    elif requestMethod == "POST":
        r = requests.post(requestUrl, headers=requestHeaders, proxies=p, verify=False, data=requestBody)
    elif requestMethod == "PUT":
        r = requests.put(requestUrl, headers=requestHeaders, proxies=p, verify=False, data=requestBody)
    elif requestMethod == "PATCH":
        r = requests.patch(requestUrl, headers=requestHeaders, proxies=p, verify=False, data=requestBody)
    elif requestMethod == "DELETE":
        r = requests.delete(requestUrl, headers=requestHeaders, proxies=p, verify=False)
    else:
        print "Unsupported Method: " + str(requestMethod)
    return r


def fixHeaders(requestHeaders):
    """Parses request headers and replaces all placeholders/variables with the associated values from the environment file."""
    newHeaders = dict()
    for key in requestHeaders:
        if "{" in requestHeaders[key]:
            value = requestHeaders[key].replace("{","").replace("}","")

            if value in env_var:
                newHeaders[key] = env_var.get(value, "")
            else:
                if value not in missing_parameters:
                    print "[!] Found header parameter not in ENVIRONMENT VARIABLES: " + str(value)
                    missing_parameters.append(value)
                newHeaders[key] = requestHeaders[key]
        else:
            newHeaders[key] = requestHeaders[key]
    return newHeaders



def doBurpStuff():
    """If BurpBudy is enabled, move requests to Intruder and Repeater."""
    if args.verbose:
        print "Checking if BurpBuddy is available..."
    # check if burpbuddy is active
    _url = BURP_BUDDY_URL + "/ping"
    try:
        _status = requests.get(_url, verify=False, timeout=3)
    except:
        _status.text = ""
    # if we get PONG as a response, API is active
    if  _status.text == "PONG":
        # pull proxy history
        if args.verbose:
            print "Getting proxy history from BurpBuddy..."
        _url = BURP_BUDDY_URL + "/proxyhistory"
        _history = requests.get(_url, verify=False)
        if _history.status_code == requests.codes.ok:
            _request_data = json.loads(_history.text)
            _repeater_url = BURP_BUDDY_URL + "/send/repeater"
            _intruder_url = BURP_BUDDY_URL + "/send/intruder"
            for _r in _request_data:
                # build burpbuddy request
                _post_data = {}
                _post_data["host"] = _r["http_service"]["host"]
                _post_data["port"] = _r["http_service"]["port"]
                _post_data["use_https"] = "true" if _r["http_service"]["protocol"] == "https" else "false"
                _post_data["request"] = _r["request"]["raw"]
                if args.verbose:
                    print _post_data
                requests.post(_repeater_url, data=json.dumps(_post_data), verify=False)
                requests.post(_intruder_url, data=json.dumps(_post_data), verify=False)


def writeAuditLog():
    """Write out a csv file with the endpoints parsed and called."""
    with open(args.output, 'wb') as csvfile:
        csvwriter = csv.writer(csvfile)
        csvwriter.writerows(audit_log)


def parseHeaders(headerFile, headerDict):
    print "\nStarting to parse headers file: " + str(headerFile)
    header_data = json.loads(open(headerFile).read())
    for value in header_data["values"]:
        #header_data[value["key"]] = str(value["value"])
        headerDict[str(value["key"])] = str(value["value"])
    if(args.verbose):
        print "---------------------------------------------------"
        print "parsed header file:  " + str(headerFile)
        print "current headers: "
        print str(headerDict)

    return headerDict


# Print splash screen
def printSplash():
    print "\n"
    print """  ___|       |     |  ___|        | |   """
    print """ |      _ \  |  _` |\___ \   _` | | __| """
    print """ |     (   | | (   |      | (   | | |   """
    print """\____|\___/ _|\__,_|_____/ \__,_|_|\__| """
    print " "
    print "Version: %s " % version
    print "\n\n"


parser = argparse.ArgumentParser(description="Cold Salt - automating API requests from common collection files.")
parser.add_argument("-a","--allmethods", help="Make API calls for all HTTP methods including state change related methods.", action="store_true", default="false")
parser.add_argument("--burpbuddy", help="IP:port of Burp Budy instance. Defaults to localhost:8081.", default="127.0.0.1:8081")
parser.add_argument("--checkonly", help="enable debug mode. No requests will be sent", action="store_true", default="false")
parser.add_argument("-c","--contenttype", help="Force or override Content-Type on requests")
parser.add_argument("-e", "--environment", help="The environment.json file to parse.")
parser.add_argument("--headers", help="The headers.json file to parse.")
parser.add_argument("-i", "--input", help="The JSON collection file, Postman or Swagger, to parse.")
parser.add_argument("--mode", choices=['curl', 'postman', 'swagger'], help="Which mode, or collection type, to use. Choices are 'curl', 'postman', or 'swagger'")
parser.add_argument("--output", help="write output/audit log to a csv")
parser.add_argument("--proxy", help="IP:Port of proxy to use. Defaults to localhost:8080.", default="127.0.0.1:8080")
parser.add_argument("-v", "--verbose", help="Enable verbose output.", action="store_true")
parser.add_argument("-x", "--xperimental", help="enable Xperimental Burp mode", action="store_true")
args = parser.parse_args()


# global variables
version = "beta 0.7.1"
safe_methods = ["GET", "HEAD", "OPTIONS"]
standard_methods = ["GET", "HEAD", "OPTIONS", "PUT", "POST", "PATCH", "DELETE", "OPTIONS"]
env_var = dict()
headers = dict()
audit_log = []
live_token = ""
requests.packages.urllib3.disable_warnings(InsecureRequestWarning) # disable this since we're sending requests through Burp
missing_parameters = []
p = {"http": "http://" + args.proxy, "https": "https://" + args.proxy}

#


printSplash()

print args.input

if args.environment:
    print "\nStarting to parse environment file: " + str(args.environment)
    env_data = json.loads(open(args.environment).read())
    for value in env_data["values"]:
        env_var[value["key"]] = value["value"]
    if(args.verbose):
        print "---------------------------------------------------"
        print "parsed environment file:  " + str(args.environment)
        print "found these environment variables: "
        print str(env_var)



# let's build a dictionary for each request and put all of them into a list
# each endpoint dictionary is {name, url, method, headers[] }
collection_items = []
globalHeaders = dict()

_file = args.input
# parse disparate input formats into simple dict entries
if args.mode == "postman":
    requestParser = CSParserPostman(verbose = args.verbose)
    collection_items = requestParser.parsePostman(args.input)
elif args.mode == "swagger":
    requestParser = CSParserSwagger(verbose = args.verbose)
    collection_items = requestParser.parseSwagger(args.input)
elif args.mode =="curl":
    requestParser = CSParserCurl(verbose = args.verbose)
    collection_items = requestParser.parseCurl(args.input)


print "\nGetting ready to send requests"

allowed_methods = []
if args.allmethods == True:
    allowed_methods = standard_methods
else:
    allowed_methods = safe_methods

# do header stuff.....
# TODO see if this step can be done when the dict is built...
globalHeaders["User-Agent"] = "coldsalt"
# see if we've got a header file to parse...
if args.headers:
    globalHeaders = parseHeaders(headerFile=args.headers, headerDict=globalHeaders)

# force Content-Type overrides other header settings
if "None" != str(args.contenttype):
    #print "overriding content type!"
    if "content-type" in headers.keys():
        globalHeaders["content-type"] = args.contenttype
    elif "Content-Type" in headers.keys():
        globalHeaders["Content-Type"] = args.contenttype


for endpoint in collection_items:
    if endpoint["method"] in allowed_methods:
        headers = dict()
        url = parseURL(endpoint["url"])
        method = endpoint["method"]
        if "body" in endpoint:
            body = parseBody(endpoint["body"])
            #print body
        else:
            body = ""

        # merge global headers
        headers = dict(fixHeaders(endpoint["headers"]).items() + globalHeaders.items())

        if(args.verbose):
            print "-------------"
            print "Request"
            print "  url: " + str(url)
            print "  method: " + str(method)
            print "  headers: " + str(headers)
            print "  body: " + str(body)
        else:
            if(args.checkonly == "false"):
                print "Sending: " + str(method) + " " + str(url)
        if(args.checkonly == "false"):
            resp = sendRequest(url, method, headers, body)

            audit_log.append([endpoint["name"], method, url, resp.status_code])
            print " - Response Status: " +  str(resp.status_code)
        else:
            audit_log.append([endpoint["name"], method, url, "Not Requested"])
    else:
        #unsafe methods
        url = parseURL(endpoint["url"])
        audit_log.append([endpoint["name"], endpoint["method"], url, "Not Requested"])

# for any parameters set in the Input file that don't have values in the Environment file, write out a stub
if len(missing_parameters) > 0:
    print "\n----------------------"

    contents = ",".join('{"key": "%s", "value": "<INSERTVALUE>"}' % param for param in missing_parameters)
    if(args.environment):
        print "Please add the following parameters to your environment file: "
        print contents
    else:
        print "Missing parameters helper file stub: "
        print '{"name": "generic config file", "values": [%s]}' % contents

if args.xperimental:
    doBurpStuff()

if(args.output):
    writeAuditLog()
